### DISCUSSION TITLE
Modbus TCP server vs Modbus wifibridge RTU

### DISCUSSION THREAD

**Discussion by Oo87** [eModbus/eModbus#350] [eModbus applications]
Hello all,

I'm using Emodbus libfor a while. I was able to setup a RTU Com to a powermeter.
I also succesfully created a bridge TCP<>RTU. My Modbus client is reading data over ethernet straight from the power meter.

The issue is that the power meter is not fully suported by the device. (basicaly, Power is reversed => 1000W should be -1000W).

The idea was to create a standalone server from Emodbus and populate and alterate the data coming from the power meter.

Unfortunately, the server isn't working as expected.

I'm now wondering if it's related to Tcp standard server while th ebridge is doing RTU over TCP.

the bridge and the server works on the same ESP32 with different port.

Both can be connected, but server isnt answering the right "data" 

Anybody can help me configure this server to act like the bridge ?

thk you all

**Comment by Miq1** [comment]
Hi there,

a bridge is not doing "RTU over TCP", but is a TCP server that happens to use a RTU client to request whatever was asked from the RTU bus.

There are two ways I see to solve your issue:

- educate your Modbus client (on a PC?) to recognize the data format of your power meter
- use a ``ModbusServerWiFi`` (or ``ModbusServerEthernet`` or ``ModbusServerETH``, whatever applies to you) and register a worker function for the function code you are going to need for the power meter as server. Create a ``ModbusClientRTU`` in the same sketch and set it up to connect to the RTU bus. In the worker function, call a ``syncRequest`` for the RTU client with the request parameters you got. Then modify the response data to fit to the requester's liking and return a response with it to your TCP requester.

The second option of course will require some work; in the end you will have a dedicated bridge just for the power meter.

**Reply by Oo87** [reply]
Hello, thank you so much for the help.

Indeed first solution is best, but my application is a inverter comunicating with power meter. So not possible.
Inverter ->modbus RTU-> gateway over TCP ->Ethernet-> esp32 ->modbus RTU-> power meter.

I tried to copy the memory from RTU client to TCP server but no succes.

```
// Modbus bridge include
#include "ModbusBridgeWiFi.h"
// Modbus RTU client include
#include "ModbusClientRTU.h"
// Modbus server include
//#include "ModbusServerRTU.h"
#include "ModbusServerWiFi.h"

const uint8_t MY_SERVER(5);
uint16_t memo[128];  
// Worker function for serverID=1, function code 0x03 or 0x04
ModbusMessage FC03(ModbusMessage request) {
  uint16_t addr = 0;        // Start address to read
  uint16_t wrds = 0;        // Number of words to read
  ModbusMessage response;

  // Get addr and words from data array. Values are MSB-first, get() will convert to binary
  request.get(2, addr);
  request.get(4, wrds);
  
  // address valid?
  if (!addr || addr > 128) {
    // No. Return error response
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
    return response;
  }

  // Modbus address is 1..n, memory address 0..n-1
  addr--;

  // Number of words valid?
  if (!wrds || (addr + wrds) > 127) {
    // No. Return error response
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
    return response;
  }

  // Prepare response
  response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(wrds * 2));

  // Loop over all words to be sent
  for (uint16_t i = 0; i < wrds; i++) {
    // Add word MSB-first to response buffer
    response.add(memo[addr + i]);
  }

  // Return the data response
  return response;
}


// Create a ModbusRTU client instance
ModbusClientRTU MB(Serial2, 4);
// Create server
ModbusBridgeWiFi MBbridge;
// Set up a Modbus server
ModbusServerWiFi MBserver;

```
In setup : 
```
  // Set RTU Modbus message timeout to 2000ms
  MB.setTimeout(2000);
  MB.onDataHandler(&handleData);

  // Start ModbusRTU background task on core 1
  MB.begin(1);
  //////////////////////////////
  ///////////////////// MBSERVER
  // Register the worker function with the Modbus server
  MBserver.registerWorker(MY_SERVER, READ_HOLD_REGISTER, &FC03);
  
  // Register the worker function again for another FC
  MBserver.registerWorker(MY_SERVER, READ_INPUT_REGISTER, &FC03);
  // Initialize server memory with consecutive values

  // Start the Modbus TCP server:
  // Port number 503, maximum of 4 clients in parallel, 10 seconds timeout
  MBserver.start(503, 4, 5000);

  //////////////////////////////
  ///////////////////// MBSERVER  

  MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE, &MB);
  // Check: print out all combinations served to Serial
  MBbridge.listServer();
  // Start the bridge. Port 502, 4 simultaneous clients allowed, 600ms inactivity to disconnect client
  MBbridge.start(port, 4, 5000);
```
In loop : 

```
    for (uint16_t i = 0; i < 128; ++i) {
      memo[i] = SDM630[i];
    }
```

I changed my fuction to : 

```
  for (uint16_t i = 0; i < wrds; i++) {
    // Add word MSB-first to response buffer
    response.add(memo[addr + i]);
  }
  response = MB.syncRequest(0, 1, READ_INPUT_REGISTER, 0, 80);
  //MB.addRequest(0, 1, READ_INPUT_REGISTER, 0, 80);
  // Return the data response
  return response;
}
```

**Reply by Miq1** [reply]
You must not use a bridge to be able to intercept the responses. A plain TCP server and a RTU client will do.

The TCP server must do a ``MB.syncRequest`` **in the** ``FC03`` worker function to wait there for the RTU server to answer. There is no need for any ``handleData`` function, as it would be asynchronous and more complicated. ``loop()`` may even be empty, as all is done in the ``FC03`` callback function.

Some random notes:
- You may have to play with the timeout times to cope with the added RTU server response and processing times.
- As this is no regular bridge anymore, you will need to add your own worker functions for any single function code you are planning to send to the power meter. There can be an ANY_FUNCTION_CODE default worker if you like, as the explicitly defined workers will take precedence.

**Comment by Miq1** [comment]
Note: I took interest in the application problem and added some functionality to the lib that may help in similar cases in the future.

If you will deploy a ``ModbusBridge`` of any type, now you can have your own filtering functions applied to all requests or responses of a given server ID. There is a modified example ``WiFi-RTU_Bridge_Filtered`` now that will showcase the functionality. Please be careful with what you are doing to the messages - these are easily rendered useless when modifying not sensibly...
